### PR TITLE
Modified subset pattern - questions inside quiz

### LINKED ISSUE #17: Fetch subsets of questions as the user moves across the quiz. Fetch all the details only for first 10 questions when fetching a quiz

### PR DESCRIPTION
**Pull request by deepansh96**
Fixes #17

### CODE REVIEW

**Comment by pritamps** [comment]
Add all the PR details. I think I'll need a call to understand this :) Will set up one for tomorrow

File: app/routers/questions.py [hunk 1]
```diff
@@ -14,3 +14,24 @@ async def get_question(question_id: str):
...
+
+    if limit:
+        pipeline.append({"$limit": limit})
+
+    if (questions := list(client.quiz.questions.aggregate(pipeline))) is not None:
+        return questions
+
+    raise HTTPException(
+        status_code=status.HTTP_404_NOT_FOUND,
+        detail=f"Questions is question set {question_set_id} not found",
```
Comment (dalmia) [inline app/routers/questions.py:36]: Questions might not be present also because the skip value might be set to some invalid. This covers all the cases IMO.

```suggestion
        detail=f"No questions not found",
```

File: app/models.py [hunk 2]
```diff
@@ -115,6 +115,7 @@ class QuestionResponse(Question):
     """Model for the response of any request that returns a question"""
 
     question_set_id: str
+    text: Optional[str] = None
```
Comment (dalmia) [inline app/models.py:118]: didn't understand why this was added.

File: app/routers/quizzes.py [hunk 3]
```diff
@@ -3,23 +3,53 @@
...
+    for question_set_index, question_set in enumerate(quiz["question_sets"]):
         questions = question_set["questions"]
-        for index, _ in enumerate(questions):
-            questions[index]["question_set_id"] = question_set["_id"]
+        for question_index, _ in enumerate(questions):
+            questions[question_index]["question_set_id"] = question_set["_id"]
 
         client.quiz.questions.insert_many(questions)
 
+        subset_with_all_details = client.quiz.questions.aggregate(
```
Comment (dalmia) [inline app/routers/quizzes.py:23]: based on the name of the variable below, this seemed more appropriate, no?
```suggestion
        subset_with_details = client.quiz.questions.aggregate(
```

File: app/tests/test_quizzes.py [hunk 4]
```diff
@@ -21,14 +40,70 @@ def test_create_quiz(self):
...
+
+        # checking the rest of the subset of questions in the quiz.
+        # They should only contain some required keys and not all the keys
+        for i in range(
+            settings.subset_size,
+            len(self.long_assessment_quiz["question_sets"][0]["questions"]),
+        ):
+            question = self.long_assessment_quiz["question_sets"][0]["questions"][i]
+
+            # for key in required_keys:
```
Comment (suryabulusu) [inline app/tests/test_quizzes.py:95]: can remove this commented out code

Comment (suryabulusu) [inline app/tests/test_quizzes.py:95]: this shouldn't be commented out na? have to assert that required keys are there. tests are passing when comments are removed

File: app/models.py [hunk 5]
```diff
@@ -130,6 +130,7 @@ class QuestionResponse(Question):
...
```
Comment (suryabulusu) [inline app/models.py:133]: why is this added? `text` is already present in `Question` model and we are inheriting it for `QuestionResponse` model

File: app/tests/test_quizzes.py [hunk 6]
```diff
@@ -21,14 +40,70 @@ def test_create_quiz(self):
...
     def test_get_quiz_returns_error_if_id_invalid(self):
         response = self.client.get(f"{quizzes.router.prefix}/00")
         assert response.status_code == 404
         response = response.json()
         assert response["detail"] == "quiz 00 not found"
+
+    def test_created_long_quiz_contains_subsets(self):
+        # the created long quiz should contain the same number of questions as the one in the provided input json
+        assert self.long_assessment_quiz_questions_length == len(
+            self.long_assessment_quiz_data["question_sets"][0]["questions"]
```
Comment (suryabulusu) [inline app/tests/test_quizzes.py:63]: a doubt -- the quizzes will work fine na if there are multiple question sets also? specifying `[0]` is only for test? 

earlier our code catered only to one question set

File: app/routers/quizzes.py [hunk 7]
```diff
@@ -3,23 +3,50 @@
...
 
         client.quiz.questions.insert_many(questions)
 
+        subset_with_details = client.quiz.questions.aggregate(
+            [
+                {"$match": {"question_set_id": question_set["_id"]}},
+                {"$limit": settings.subset_size},
+            ]
+        )
+
```
Comment (sudheshna-donthineni) [inline app/routers/quizzes.py:29]: What is happening here?
